### PR TITLE
Fix asking for username/password in Publisher

### LINKED ISSUE #4790: 'IO' object has no attribute 'ask'
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: Ubuntu 20.04
- **Poetry version**: 1.2.0a2
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: 

## Issue
After adding a repository but no credentials, I cannot publish:
```
> poetry config repositories.xyz.url "http://python.repo/an/index/"
```
Fix https://github.com/python-poetry/poetry/issues/3104#issuecomment-972160123, then
```
> poetry publish -r xyz -vvv
Loading configuration file /home/user/.config/pypoetry/config.toml
Loading configuration file /home/user/.config/pypoetry/auth.toml
Loading configuration file /home/user/project/poetry.toml
Adding repository xyz (http://python.repo/an/index/)


  Stack trace:

  8  ~/.local/share/pypoetry/venv/lib/python3.8/site-packages/cleo/application.py:330 in run
      328│ 
      329│             try:
    → 330│                 exit_code = self._run(io)
      331│             except Exception as e:
      332│                 if not self._catch_exceptions:

  7  ~/.local/share/pypoetry/venv/lib/python3.8/site-packages/poetry/console/application.py:180 in _run
      178│         self._load_plugins(io)
      179│ 
    → 180│         return super()._run(io)
      181│ 
      182│     def _configure_io(self, io: IO) -> None:

  6  ~/.local/share/pypoetry/venv/lib/python3.8/site-packages/cleo/application.py:425 in _run
      423│                 io.set_input(ArgvInput(argv))
      424│ 
    → 425│         exit_code = self._run_command(command, io)
      426│         self._running_command = None
      427│ 

  5  ~/.local/share/pypoetry/venv/lib/python3.8/site-packages/cleo/application.py:467 in _run_command
      465│ 
      466│         if error is not None:
    → 467│             raise error
      468│ 
      469│         return event.exit_code

  4  ~/.local/share/pypoetry/venv/lib/python3.8/site-packages/cleo/application.py:451 in _run_command
      449│ 
      450│             if event.command_should_run():
    → 451│                 exit_code = command.run(io)
      452│             else:
      453│                 exit_code = ConsoleCommandEvent.RETURN_CODE_DISABLED

  3  ~/.local/share/pypoetry/venv/lib/python3.8/site-packages/cleo/commands/base_command.py:118 in run
      116│         io.input.validate()
      117│ 
    → 118│         status_code = self.execute(io)
      119│ 
      120│         if status_code is None:

  2  ~/.local/share/pypoetry/venv/lib/python3.8/site-packages/cleo/commands/command.py:85 in execute
       83│ 
       84│         try:
    →  85│             return self.handle()
       86│         except KeyboardInterrupt:
       87│             return 1

  1  ~/.local/share/pypoetry/venv/lib/python3.8/site-packages/poetry/console/commands/publish.py:78 in handle
       76│         )
       77│ 
    →  78│         publisher.publish(
       79│             self.option("repository"),
       80│             self.option("username"),

  AttributeError

  'IO' object has no attribute 'ask'

  at ~/.local/share/pypoetry/venv/lib/python3.8/site-packages/poetry/publishing/publisher.py:82 in publish
       78│         )
       79│         # Requesting missing credentials but only if there is not a client cert defined.
       80│         if not resolved_client_cert:
       81│             if username is None:
    →  82│                 username = self._io.ask("Username:")
       83│ 
       84│             # skip password input if no username is provided, assume unauthenticated
       85│             if username and password is None:
       86│                 password = self._io.ask_hidden("Password:")
```

### PR DESCRIPTION
**Pull request by NiklasRosenstein**
Fixes #4790

Somewhere between Cleo 0.8.0 and 1.0.0a4, the `IO.ask()` and `IO.ask_hidden()` methods were removed. I added `Publisher.ask()` which is almost equivalent to Cleo's `Command.ask()`, only that it also accepts a `hidden` argument.

### CODE REVIEW

File: src/poetry/publishing/publisher.py
Comment (neersighted) [inline src/poetry/publishing/publisher.py:117]: I'm a little concerned that though this code path isn't used, the fact that passing in a `Question` instance will ignore the `hidden` parameter could be an issue in the future. Why don't we just create the question inline instead of using this helper?

**Review by neersighted — CHANGES_REQUESTED** [review]

**Comment by NiklasRosenstein** [comment]
Hey @neersighted , what do you think of it now?

**Comment by NiklasRosenstein** [comment]
Pinging @radoering as the last committer on Poetry hoping to get some momentum in here. 🙃

**Comment by radoering** [comment]
Two remarks:

1. Is there a reason why the import of `Question` is local?
2. I wonder if it is feasable to write a test in order to detect possible API changes in the future?

**Comment by neersighted** [comment]
> Hey @neersighted , what do you think of it now?

Sorry, fell down the work hole -- I think @radoering should be able to get you over the finish line :)

**Comment by branchvincent** [comment]
@NiklasRosenstein sorry for the long back and forth, are you still interested in finishing this out with a test? You could follow the examples using `inputs` from [test_init.py](https://github.com/python-poetry/poetry/blob/master/tests/console/commands/test_init.py)

No worries if not, I can help with this last step

**Comment by dimbleby** [comment]
Just spotted this after #5889 was merged.  In that MR I removed this code altogether - which certainly fixes the bug that you were hitting.

Do you actually care about asking for username and password here?  IMO if this is to happen it would be better on the `Command`, but deleting the code seems satisfactory to me, the command line accepts `--password` and `--username` already and I don't see any compelling reason to treat just these two parameters on just this one command exceptionally.

**Comment by neersighted** [comment]
> Just spotted this after #5889 was merged. In that MR I removed this code altogether - which certainly fixes the bug that you were hitting.
> 
> Do you actually care about asking for username and password here? IMO if this is to happen it would be better on the `Command`, but deleting the code seems satisfactory to me, the command line accepts `--password` and `--username` already and I don't see any compelling reason to treat just these two parameters on just this one command exceptionally.
> 
> (Also this has been broken for so long that it's clear that no-one has been using the function!)

I merged that PR in light of this one existing / planning to circle back to it. I do think that the interactive credential support is an anti-pattern and that given we've greatly improved support for non-interactive credentials there is no more need for it. Combine that with adaptation to newer Cleo versions being kludgy and I don't particular care to resurrect it.

That being said, @NiklasRosenstein if you disagree, please feel free to post a new version of this patch/re-open this PR, especially if you can come up with a more elegant design. @dimbleby's input looks good to me.

**Pull request closed without merging**
